### ISSUE TITLE
BLinkSwap

### ISSUE THREAD

**Issue by SomethingGettingWrong** [loki-project/loki-improvement-proposals#15]
***Here is a way to swap to other coins easily and gives you a reason to use the other mining emission and give it to service nodes***

 Basically all you have to do is have the quorum take a transaction blink it but keep it locked in the memory pool for a given period  of blocks...equal to the alt coins or btc block confirmation time... while also overriding the normal Loki chains mempool at the quorum level l!!  every block! 

 if the other chains transaction doesnt go through then you unlock the blink transaction and the transaction never took place... THERS NO NEED on an atomic swap... to even move the funds if the other coins mempool tx doesnt go through and confirm a given number of times set in the loki wallet...

 we just need service nodes to  sync other coins .. for monitoring only and participate in an elevated "blink coded" atomic swap transaction and burn loki as an elevated fee over blink.. and also pay the service nodes who sync other coins for monitoring teh mining emission when you swithc to pure POS mining.

Now before you say oh wait what if they send the btc.... and we dont blink the loki..... well thats why you blink swap the Loki first... before they send the btc etc..  if you dont blink loki then they dont send btc etc.

check pointing with POW/POS presents an issue but if it goes to pure POS then this issue dissapears completely

**Comment by Lucifer1903** [comment]
I like this idea, definitely think it should be persued after pulse is implemented.

**Comment by KeeJef** [comment]
So the basic idea here i like, we could get each service node to sync all block headers of the bitcoin chain and reference some semi trusted SPV nodes for confirmation that a bitcoin transaction had occurred to a specific address. We would want the Service Node quourms who did this to be large, like 200 Nodes, since that would allow for swaps where the punishment for lying would be the destroyal of staked funds which in the case that 200 nodes lied would be over a one million dollars which secures swaps under that amount. 

And on the Loki side we could allow specially tagged Loki txes to be frozen while waiting for a set of conditions to be satisfied on the Bitcoin side of things. 

If we wanted to create a more orderbooky scenario we could allow users to burn Loki to the orderbook which could later be minted and sent to a specific address if the 200 node quorum signed off on it 

However the issues come with the practicality of this scheme 

**Bitcoin Wallet Side issues** 

If we assume that this BTC to Loki swap is made of some users who want to swap BTC and some users who want to swap Loki, a peer to peer type system, then we will have many people who want to swap only small amounts at a desired exchange rate. 

The issue comes with swaps from BTC to Loki, where a Bitcoin transaction must be constructed with the users wallet. If the user only wants to consume one of the orders on the Loki side then they must only send one transaction to X desired address. This is simple enough, most users understand how to send a transaction to an address. And this would require no modification to any existing BTC wallet. 

However the more realistic case is that there are thousands of orders in the order book, as there is with traditional exchanges. And when a user wants to execute a swap they want to consume multiple proposed swaps. However in this sense if the user is swapping from BTC to Loki it becomes very tricky, since the user will need to construct a transaction which sends to multiple bitcoin addresses to satisfy the Service Nodes that all users on the Loki side have been paid. 

Most consumer grade bitcoin wallets don't allow users to easily construct transactions with multiple payout addresses. Which makes this difficult. 

On the Loki side since we fully control the wallet software, when Loki need to be converted into BTC its easy because the Loki wallet software can easily support multiple transactions wherein multiple destinations are paid.

**Partially filled orders**

We also have an issue with partially filled orders, for example if the user wants to buy 400 Loki but there is orders on the books for 100 and 350 Loki, then the user must buy 450 Loki or 100 Loki, there might be a potential way to pre split outputs as they are sent to make this more granular, however i'm not aware of how to do this.

Note these thoughts are unorganized i might look into this further, but the bitcoin wallet thing seems like an issue considering most people use Coinbase

**Comment by Lucifer1903** [comment]
@KeeJef 
Service nodes could have people lock their Loki in an order book and people who want to fill the orders would have to request a single address from the service node quorum which the quorum controls. The service nodes provide the single address, bitcoin is sent there and then the service nodes sends it to the respective orders that are being filed and then they send the loki to the respective order filler.

**Comment by SomethingGettingWrong** [comment]
If the service node actually syncs the blockchain and is trusted or etc. Then realisticly loki wallet could generate a btc wallet as well to be used one time for the transfer no?  At that point the loki wallets could communicate to the service nodes who would be monitoring the chain. 


You always send the LOKI first  since the servicenodes have it locked (via concensus) until the btc arrives. If it doesnt it is released. 

IF someone sends BTC first  to the persons address.... then their is nothing you can really do about that as they got scammed. Loki always needs to be locked before other altcoin funds are sent. then it doesnt matter.

**Comment by SomethingGettingWrong** [comment]
I wast even thinking of doing an order book scenario... 

I was thinking more of say.. session has BTC wallet and LOKI wallet... thats remote noded to whichever BTC and LOKI Servicenode you trust. 

You send someone a LOKI  address  that your session (loki wallet) created and they send you a BTC ADDRESS their loki wallet created. Paper btc wallet. 

THe loki gets sent first... You both see loki in the mempool held by servicenodes....   if he chooses not to send btc... Then the loki will unlock after so many blocks back into the original wallet. Etc. 

Once he sends it the service nodes wait for confirmation of a set amount and when say 20 mins has passed loki and btc are both unlocked in the wallet.

**Comment by William-E-Coyote** [comment]
Is this a good idea for a SNApp? There are many amazing ideas, but the responsibility of the foundation and LOKI dev team are to produce the LOKI.net, sessions, and other core capabilities. It seems to me that the SNApp architecture is rather flexible and that a good designer or team of designers could develop this SNApp and charge a bit for their work.

**Comment by KeeJef** [comment]
@William-E-Coyote No, this would need to operate at a lower level than a SNApp, since this would require the Service Node run some form of bitcoin client

@Lucifer1903 That would work, but i feel uncomfortable with the Service Node network having custody of bitcoin, and if we had a de-registration event we could loose the nodes involved in that multisig address and those coins could no longer be moved

**Comment by SomethingGettingWrong** [comment]
The method I'm talking about wouldnt require the service nodes to have custody of either loki nor Bitcoin. It doesn't need to be a snap are require bisq the core concept revolves around just sending loki first and the quorom freezing for a period of time until the Bitcoin is in the mempool of the 
itcoin network where it will be confirmed if btc is confirmed are the service nodes releasing the funds back into the loki wallet. 

This relies on blink just to freeze the funds for clarification until the other party's btc enters the mempool and confirms. If it never enters and confirms then loki us released.

**Comment by KeeJef** [comment]
@SomethingGettingWrong i understand what you are saying, i just dont think it would be very useful, since people are used to the experience of using an exchange, they don't expect to have to find someone who has the exact amount of Loki that they want to swap for BTC. 

This feature would be a lot of work to deploy and if its going to be minimally used then theres not much point spending lots of dev time on it. If you can think of a way to extend this to have an orderbook and resolve the issues with non supporting BTC wallets then i think it would be a good idea.

**Comment by Lucifer1903** [comment]
@KeeJef how about locking loki in an orderbook using BlinkSwap and having a SNApp where you enter how much loki you want to buy, the SNApp would then calculate from the order book the price for that amount of loki. If you wanted to accept that price you would have to pay a small deposit relative to the amount you want to buy (let's say 1%) and your loki receiving address. When that deposits is paid you have 2 hours to send the bitcoin payment to the address the SNApp provides.
The bitcoin payment address would be a multisig address created by 50 Service Nodes at the time of deposit payment and would require 25 signatures to sign transactions. When the bitcoin payment is sent to the address the Service Nodes create a transaction to redistribute the payment between all the orders used to create the BlinkSwap transaction. This redistribution transaction can be created and sent as soon as the payment transaction is in the mempool (bitcoin consensus rules transactions can spend outputs from other transactions in the same block as long as they are at an earlier position) but the redistribution transaction should also be shared with the rest of the Service Node network along with the loki receiving address for redundancy until the redistribution transaction has 10 blocks and the orders are released to the Loki receiving address.

This method has a orderbook (even if it is only sell orders) which allows easy price verification for a set amount of loki, and solves the non-supporting BTC wallets issue while also allowing quick processing of transactions reducing the danger time where mass deregistration could cause greater than 50% of the Bitcoin quorum SNs to go down, the danger can also be reduced by further increasing the quorum size or reducing the payment time from 2 hours to 1 hour.

That being said this would be a lot of work and there are definitely other places where dev time would be better spent for the time being.

**Comment by KeeJef** [comment]
@Lucifer1903 i don't think this really solves the issue, and it also introduces the issue of the Bitcoin chain being reorganized, the Service Nodes would want to wait for 6 confirmations on the BTC chain before releasing Loki.

**Comment by Lucifer1903** [comment]
@KeeJef sorry could you explain which part it doesn't solve? Is it that there's still a possibility that mass deregistration could cause greater than 50% of the Quorum to be deregistered?

Also I did say to wait for 10 blocks (I meant confirmations) but you're right that can be reduced to 6.

**Comment by KeeJef** [comment]
Yes, any method of custody where the Service Nodes are given control over a significant amount of BTC in a multisig wallet is prone to de-registration events, or even just temporary communications fallouts, i think its fairly weak to have a custody based solution in this case. I think the only way to do this effectively on the BTC side would be to have the wallet send a transfer to multiple parties in the same TX, which is possible, lots of wallets implement this feature, like the Electrum,  Bitcoin core and Trezor wallets, however most mobile wallets and Ledger do not have support. I think this would lead to low adoption.

**Comment by Lucifer1903** [comment]
Makes sense

**Comment by SomethingGettingWrong** [comment]
> 
> 
> @SomethingGettingWrong i understand what you are saying, i just dont think it would be very useful, since people are used to the experience of using an exchange, they don't expect to have to find someone who has the exact amount of Loki that they want to swap for BTC.
> 
> This feature would be a lot of work to deploy and if its going to be minimally used then theres not much point spending lots of dev time on it. If you can think of a way to extend this to have an orderbook and resolve the issues with non supporting BTC wallets then i think it would be a good idea.

You could burn an amount of loki in a tx (with extra information in the header that specified the amoutn of loki and bitcoin you would accept for given loki) and a session ID  generated specifcily for the transaction. 

Then loki blocks could be used to pull API information of current "orders"  which would be session ID's tied to it ... you open it up and wait for the transfer.... after messaging them... YOU DONT EVEN HAVE TO KNOW WHO THEY ARE!!!

Add an api to lokiblocks

curl -w "\n" -X GET "www.lokiblocks.com/api/transaction/sessionblinkswap

That you can browse in session with... see Id's that are paired up and price in btc (in usd value) and Loki ( in usd value)  click on it  and wait for them to swap ... by messaging them.
